### PR TITLE
Model builder pattern

### PR DESCRIPTION
**Pull request by lixitrixi**
This is a continuation of work from #16. Rebasing has proven difficult so I have based this branch on the current merged changes and manually copied my work over.

### CODE REVIEW

File: src/common/ast/ast.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#[derive(Debug, PartialEq)]
+pub struct DecisionVariable {
+    pub domain: Domain,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Domain {
+    BoolDomain,
+    IntDomain(Vec<Range<i32>>),
```
Comment (ChrisJefferson) [inline src/common/ast/ast.rs:32]: This is fine for now, but at some point someone will have to write (or find) a lovely library for ranges of ranges -- at the moment the domains [0..3), [3..6), [0..3], [1..6) and [0..6) would all be different, despite them having the same underlying values. I'm sure at a later date there will be a demand for interesting + unioning such domains as well.

File: src/common/ast/ast.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Range<A> {
+    Single(A),
+    Bounded(A, A),
```
Comment (ChrisJefferson) [inline src/common/ast/ast.rs:38]: Minor issue, make clear if Bounded(2,5) is {2,3,4,5} or {2,3,4} (I've seen both used).

File: src/common/ast/model_builder.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        }
+    }
+
+    pub fn add_constraint(mut self, constraint: Expression) -> Self {
+        self.constraints.push(constraint);
+        self
+    }
+
+    pub fn add_var(mut self, name: Name, domain: Domain) -> Self {
+        self.variables.insert(
```
Comment (ChrisJefferson) [inline src/common/ast/model_builder.rs:23]: as above, for now maybe just `assert!(self.variables.get(name).is_none())` ?

File: src/common/ast/ast.rs (hunk 2 above)
Comment (lixitrixi) [reply src/common/ast/ast.rs:38]: This is a good point, @ozgurakgun what are your thoughts?

Comment (ozgurakgun) [reply src/common/ast/ast.rs:38]: both sides inclusive was my thinking

File: src/common/ast/ast.rs (hunk 1 above)
Comment (ozgurakgun) [reply src/common/ast/ast.rs:32]: fully agreed

**Comment by ozgurakgun** [comment]
can you try merging main onto this pr @lixitrixi ? I tried to fix the rustfmt github action by using a fixed version number, just want to see whether that worked.

File: src/common/ast/ast.rs (hunk 2 above)
Comment (lixitrixi) [reply src/common/ast/ast.rs:38]: Should I rename it to `BoundedInclusive` then?

Comment (ozgurakgun) [reply src/common/ast/ast.rs:38]: There is only one bounded, so I think the name is ok. We want it documented though, definitely.

**Comment by lixitrixi** [comment]
I have removed the `add_var_str` method for now since it's just a less verbose in application version of the regular `add_var` method. Perhaps if this builder will be used for constructing models by hand we could have a "find" method which accepts a string, but it seems like we're skipping that and going straight to parsing from structured input.

**Comment by ozgurakgun** [comment]
are we abandoninig this one or do you want to update it @lixitrixi ?

**Comment by lixitrixi** [comment]
@ozgurakgun Was the code structure PR passing rustfmt before you merged it?

**Comment by ozgurakgun** [comment]
sure!

if it was me I would write the astjson parser and see what kind of builder makes sense for that.

**Comment by lixitrixi** [comment]
> sure!
> 
> if it was me I would write the astjson parser and see what kind of builder makes sense for that.

That makes the most sense... I will close this and implement necessary model methods over on #32 as I work.

**Pull request closed without merging**